File: spec/support/components/projects/top_menu.rb
# frozen_string_literal: true

#-- copyright
# OpenProject is an open source project management software.
# Copyright (C) the OpenProject GmbH
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License version 3.
#
# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
# Copyright (C) 2006-2013 Jean-Philippe Lang
# Copyright (C) 2010-2013 the ChiliProject Team
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
#
# See COPYRIGHT and LICENSE files for more details.
#++
require "support/finders/test_selector_finders"

module Components
  module Projects
    class TopMenu
      include Capybara::DSL
      include Capybara::RSpecMatchers
      include RSpec::Matchers
      include ::TestSelectorFinders

      def toggle
        page.find_by_id("projects-menu").click
        wait_for_network_idle(timeout: 10)
      end

      # Ensures modal registers as #open? before proceeding
      def toggle!
        toggle
        expect_open
      end

      def open?
        page.has_selector?(search_selector)
      end

      def switch_mode(mode)
        within_test_selector("op-header-project-select") do
          find_button(mode).click
        end
      end

      def expect_current_mode(mode)
        within_test_selector("op-header-project-select") do
          expect(page).to have_css(".SegmentedControl-item--selected", text: mode)
        end
      end

      def expect_current_project(name)
        page.find_by_id("projects-menu", text: name)
      end

      def expect_open
        page.find(search_selector)
      end

      def expect_closed
        expect(page).to have_no_selector(search_selector)
      end

      def search(query)
        search_field.set query
      end

      def clear_search
        search_field.set ""
        search_field.send_keys :backspace
      end

      def search_and_select(query)
        search query
        wait_for_network_idle
        selector = "#{results_selector} #{item_selector}"
        item = page.first(selector, text: query, wait: 5) || page.find(selector, wait: 5)
        item.click
      end

      def search_results
        page.find results_selector, wait: 10
      end

      def search_field
        page.find search_selector, wait: 10
      end

      def expand_node_for(name)
        item = page.find("#{results_selector} #{item_selector}", text: name, wait: 10)
        item.find(:xpath, "preceding-sibling::*[contains(@class, 'TreeViewItemToggle')]").click
      end

      def expect_result(name, disabled: false, workspace_badge: nil)
        selector = disabled ? item_disabled_selector : item_selector
        item = page.find("#{results_selector} #{selector}", text: name, wait: 10)

        return if workspace_badge.nil?

        if workspace_badge
          expect(item).to have_octicon
          expect(item).to have_primer_text(workspace_badge, class: "description")
        else
          expect(item).to have_no_octicon
          expect(item).to have_no_primer_text(class: "description")
        end
      end

      def expect_no_result(name)
        expect(page).to have_no_selector("#{results_selector} #{item_selector}", text: name, wait: 5)
      end

      def expect_blankslate
        expect(page).not_to have_test_selector("op-header-project-select--no-results", wait: 0)
      end

      def expect_item_with_hierarchy_level(hierarchy_level:, item_name:)
        hierarchy_selector = ".TreeViewItemContainer[style*='--level: #{hierarchy_level};']"
        expect(page)
          .to have_css("#{results_selector} #{hierarchy_selector} #{item_selector}", text: item_name, wait: 10)
      end

      def expect_project_create_button
        expect(page).to have_test_selector("create-project-btn")
      end

      def expect_no_project_create_button
        expect(page).to have_no_test_selector("create-project-btn")
      end

      def expect_project_list_button
        expect(page).to have_test_selector("list-project-btn")
      end

      def expect_no_project_list_button
        expect(page).to have_no_test_selector("list-project-btn")
      end

      def item_selector
        '[data-test-selector="op-header-project-select--item"]'
      end

      def item_disabled_selector
        "#{item_selector}[aria-disabled='true']"
      end

      def results_selector
        '[data-test-selector="op-header-project-select--list"]'
      end

      def active_item_selector
        "#{item_selector}[aria-current='true']"
      end

      def remove_item_selector
        "[data-test-selector='op-header-project-select--remove-item']"
      end

      def search_selector
        "[data-test-selector='op-header-project-select--search']"
      end
    end
  end
end

File: frontend/src/app/core/setup/globals/global-listeners/top-menu.ts
// -- copyright
// OpenProject is an open source project management software.
// Copyright (C) 2012-2022 the OpenProject GmbH
//
// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License version 3.
//
// OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
// Copyright (C) 2006-2013 Jean-Philippe Lang
// Copyright (C) 2010-2013 the ChiliProject Team
//
// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License
// as published by the Free Software Foundation; either version 2
// of the License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
//
// See COPYRIGHT and LICENSE files for more details.
//++
import { findAllFocusableElementsWithin } from 'core-app/shared/helpers/focus-helpers';

export const ANIMATION_RATE_MS = 100;

export class TopMenu {
  private hover = false;

  private menuIsOpen = false;

  constructor(readonly menuContainer:JQuery) {
    this.withHeadingFoldOutAtBorder();
    this.setupDropdownClick();
    this.registerEventHandlers();
    this.closeOnBodyClick();
    this.accessibility();
    this.skipContentClickListener();
  }

  skipContentClickListener():void {
    // Skip menu on content
    jQuery('#skip-navigation--content').on('click', () => {
      // Skip to the breadcrumb or the first link in the toolbar or the first link in the content (homescreen)
      const selectors = '.first-breadcrumb-element a, .toolbar-container a:first-of-type, #content a:first-of-type';
      const visibleLink = jQuery(selectors)
        .not(':hidden')
        .first();

      if (visibleLink.length) {
        visibleLink.trigger('focus');
      }
    });
  }

  accessibility():void {
    jQuery('.op-app-menu--dropdown').attr('aria-expanded', 'false');
  }

  toggleClick(dropdown:JQuery):void {
    if (this.menuIsOpen) {
      if (this.isOpen(dropdown)) {
        this.closing();
      } else {
        this.open(dropdown);
      }
    } else {
      this.opening();
      this.open(dropdown);
    }
  }

  // somebody opens the menu via click, hover possible afterwards
  opening():void {
    this.startHover();
    this.menuIsOpen = true;
    this.menuContainer.trigger('openedMenu', this.menuContainer);
  }

  // the entire menu gets closed, no hover possible afterwards
  closing():void {
    this.stopHover();
    this.closeAllItems();
    this.menuIsOpen = false;
    this.menuContainer.trigger('closedMenu', this.menuContainer);
  }

  stopHover():void {
    this.hover = false;
    this.menuContainer.removeClass('hover');
  }

  startHover():void {
    this.hover = true;
    this.menuContainer.addClass('hover');
  }

  closeAllItems():void {
    this.openDropdowns().each((ix, item) => {
      this.close(jQuery(item));
    });
  }

  closeOnBodyClick():void {
    const wrapper = document.getElementById('wrapper');
    if (!wrapper) {
      return;
    }

    wrapper.addEventListener('click', (evt) => {
      if (this.menuIsOpen && !this.openDropdowns()[0].contains(evt.target as HTMLElement)) {
        this.closing();
      }
    }, true);
  }

  openDropdowns() {
    return this.menuContainer.find('.op-app-menu--item_dropdown-open');
  }

  dropdowns() {
    return this.menuContainer.find('.op-app-menu--item_has-dropdown');
  }

  withHeadingFoldOutAtBorder():void {
    let menuStartPosition;
    const next = this.menuContainer.next();
    const wikiHeading = this.menuContainer.next().children().next().first();
    if (next.get(0)?.tagName === 'H2') {
      menuStartPosition = this.menuContainer.next().innerHeight()! + this.menuContainer.next().position().top;
      this.menuContainer.find('.op-app-menu--body').css({ top: menuStartPosition });
    } else if (this.menuContainer.next().hasClass('wiki-content')
      && wikiHeading.get(0)?.tagName === 'H1') {
      menuStartPosition = wikiHeading.innerHeight()! + wikiHeading.position().top;
      this.menuContainer.find('.op-app-menu--body').css({ top: menuStartPosition });
    }
  }

  setupDropdownClick():void {
    this.dropdowns().each((ix, it) => {
      jQuery(it).find('.op-app-menu--item-action').click((e) => {
        this.toggleClick(jQuery(it));
        e.preventDefault();
      });
    });
  }

  isOpen(dropdown:JQuery):boolean {
    return dropdown.filter('.op-app-menu--item_dropdown-open').length === 1;
  }

  isClosed(dropdown:JQuery):boolean {
    return !this.isOpen(dropdown);
  }

  open(dropdown:JQuery):void {
    this.dontCloseWhenUsing(dropdown);
    this.closeOtherItems(dropdown);
    this.slideAndFocus(dropdown, () => {
      dropdown.trigger('opened', dropdown);
    });
  }

  close(dropdown:JQuery, immediate?:boolean):void {
    this.slideUp(dropdown, !!immediate);
    dropdown.trigger('closed', dropdown);
    this.removeStoppingOfEventPropagation(dropdown);
  }

  closeOtherItems(dropdown:JQuery):void {
    this.openDropdowns().each((ix, it) => {
      if (jQuery(it) !== jQuery(dropdown)) {
        this.close(jQuery(it), true);
      }
    });
  }

  dontCloseWhenUsing(dropdown:JQuery):void {
    setTimeout(() => {
      jQuery(dropdown).find('li').click((event) => {
        event.stopPropagation();
      });

      jQuery(dropdown).bind('mousedown mouseup click', (event) => {
        event.stopPropagation();
      });
    }, 0);
  }

  removeStoppingOfEventPropagation(dropdown:JQuery):void {
    jQuery(dropdown).find('li').unbind('click');
    jQuery(dropdown).unbind('mousedown mouseup click');
  }

  slideAndFocus(dropdown:JQuery, callback:any) {
    this.slideDown(dropdown, callback);
    setTimeout(() => this.focusFirstInputOrLink(dropdown), ANIMATION_RATE_MS);
  }

  slideDown(dropdown:JQuery, callback:any) {
    const toDrop = dropdown.find('.op-app-menu--dropdown');
    dropdown.addClass('op-app-menu--item_dropdown-open');
    toDrop.slideDown(ANIMATION_RATE_MS, callback).attr('aria-expanded', 'true');
  }

  slideUp(dropdown:JQuery, immediate:boolean):void {
    const toDrop = jQuery(dropdown).find('.op-app-menu--dropdown');
    dropdown.removeClass('op-app-menu--item_dropdown-open');

    if (immediate) {
      toDrop.hide();
    } else {
      toDrop.slideUp(ANIMATION_RATE_MS);
    }

    toDrop.attr('aria-expanded', 'false');
  }

  // If there is ANY input, it will have precedence over links,
  // i.e. links will only get focused, if there is NO input whatsoever
  focusFirstInputOrLink(dropdown:JQuery):void {
    const focusable = findAllFocusableElementsWithin(dropdown.find('.op-app-menu--dropdown')[0]);
    const toFocus = focusable[0] as HTMLElement;
    if (!toFocus) {
      return;
    }
    // actually a simple focus should be enough.
    // The rest is only there to work around a rendering bug in webkit (as of Oct 2011),
    // occurring mostly inside the login/signup dropdown.
    toFocus.blur();
    setTimeout(() => {
      toFocus.focus();
    }, 10);
  }

  registerEventHandlers():void {
    const toggler = jQuery('#main-menu-toggle');

    this.menuContainer.on('closeDropDown', (event:Event) => {
      this.close(jQuery(event.target as HTMLElement));
    }).on('openDropDown', (event) => {
      this.open(jQuery(event.target));
    }).on('closeMenu', () => {
      this.closing();
    }).on('openMenu', () => {
      this.open(this.dropdowns().first());
      this.opening();
    });

    toggler.on('click', () => { // click on hamburger icon is closing other menu
      this.closing();
    });
  }
}
